core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+                try:
+                    return (list([
+                        exp_services.get_exploration_from_model(
+                            exp_model,
+                            pre_v4_states_conversion_func=(
+                                cls._force_rule_spec_subjects_to_answer))
+                        for exp_model in exp_models_by_versions
+                    ]), False, None)
+                except utils.ExplorationConversionError as error:
+                    return (None, True, error)
+            return (None, True, error)
+
+    @classmethod
+    def _get_all_exploration_versions(cls, exp_id, max_version):
+        """Return all exploration versions for the given exploration ID in
+        descending order by version, or an empty list if none could be found
+        (such as if the exploration was permanently deleted). This function can
+        retrieve deleted explorations. Returns None if any exploration that
+        would have been returned failed a migration.
+        """
+        # NOTE(bhenning): It's possible some of these answers were submitted
+        # during a playthrough where the exploration was changed midway. There's
+        # not a lot that can be done about this since it's not possible to
+        # demultiplex the stream of answers and identify which session they are
+        # associated with.
+        versions = list(reversed(xrange(1, max_version + 1)))
+        exp_models_by_versions, migration_error = (
+            cls._get_exploration_models_by_versions(exp_id, versions))
+        if not exp_models_by_versions:
+            return ([], migration_error)
+        exps, failed_migration, migration_error = (
+            cls._get_explorations_from_models(exp_models_by_versions))
+        return (None, migration_error) if failed_migration else (exps, None)
+
+    # This function comes from extensions.answer_summarizers.models.
+    @classmethod
+    def _get_hashable_value(cls, value):
+        """This function returns a hashable version of the input value. If the
+        value itself is hashable, it simply returns that value. If it's a list,
+        it will return a tuple with all of the list's elements converted to
+        hashable types. If it's a dictionary, it will first convert it to a list
+        of pairs, where the key and value of the pair are converted to hashable
+        types, then it will convert this list as any other list would be
+        converted.
+        """
+        if isinstance(value, list):
+            return tuple([cls._get_hashable_value(elem) for elem in value])
+        elif isinstance(value, dict):
+            return cls._get_hashable_value(
+                [(cls._get_hashable_value(key), cls._get_hashable_value(value))
+                 for (key, value) in value.iteritems()])
+        else:
+            return value
+
+    @classmethod
+    def _permute_index(cls, value_list, idx):
```

Comment:
We're also slightly confused when this is hit (looking over rules.json and the logic again, it doesn't seem like this nested permuting is needed since complex structures seem to be dicts of dicts, rather than dicts of lists). However, deciding to keep it since despite it being O(n!!) the nested factorial only happens in cases where we need this algorithm, which should be sufficiently rare that this probably won't impact the overall performance of the job. We still need the top-level factorial since we need to permute the rule spec params themselves.